KAFKA-14274 [2-5/7]: Introduction of more infrastructure for forthcoming fetch request manager

kirktrue opened the pull request:
This continues the work of providing the groundwork for the [fetch refactoring work](https://issues.apache.org/jira/browse/KAFKA-14274) by introducing some new classes and refactoring the existing code to use the new classes where applicable.

Changes:

* Minor clean up of the events classes to make data immutable, private, and implement `toString()`.
* Added `IdempotentCloser` which prevents a resource from being closed more than once. It's general enough that it could be used elsewhere in the project, but it's limited to the consumer internals for now.
* Extract inner classes from `AbstractFetch`  into top-level classes to a) buffer raw results (`FetchBuffer`), and b) collect raw results into `ConsumerRecords` (`FetchCollector`). These can be tested and changed in isolation from the core fetcher logic.
* Added `NodeStatusDetector` which abstracts methods from `ConsumerNetworkClient` so that it and `NetworkClientDelegate` can be used in `AbstractFetch` via the interface instead of using `ConsumerNetworkClient` directly.

kirktrue committed:
KAFKA-14274: Introduction of more infrastructure for forthcoming fetch request manager

This continues the work of providing the groundwork for the fetch
refactoring work by introducing some new classes and refactoring the
existing code to use the new classes where applicable.

Changes:

* Minor clean up of the events classes to make data immutable,
  private, and implement toString().
* Added IdempotentCloser which prevents a resource from being closed
  more than once. It's general enough that it could be used elsewhere
  in the project, but it's limited to the consumer internals for now.
* Split core Fetcher code into classes to buffer raw results
  (FetchBuffer) and to collect raw results into ConsumerRecords
  (FetchCollector). These can be tested and changed in isolation from
  the core fetcher logic.
* Added NodeStatusDetector which abstracts methods from
  ConsumerNetworkClient so that it and NetworkClientDelegate can be
  used in AbstractFetch via the interface instead of using
  ConsumerNetworkClient directly.

kirktrue commented:
Test failures are unrelated:

- `kafka.api.ConsumerBounceTest.testClose()`
- `kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestIncludingClusterAuthorizedOperations()`: KAFKA-15419
- `o.a.k.controller.QuorumControllerTest.testBalancePartitionLeaders()`: KAFKA-15052
- `o.a.k.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManagerTest.testRemoteLogSizeCalculationWithSegmentsOfTheSameEpoch()`

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -0,0 +1,372 @@
+                final CompletedFetch nextInLineFetch = fetchBuffer.nextInLineFetch();
+
+                if (nextInLineFetch == null || nextInLineFetch.isConsumed()) {
+                    final CompletedFetch records = fetchBuffer.peek();
```
This is an existing issue. But would it be better to name `records` `completedFetch`?

kirktrue replied:
Will do.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -0,0 +1,372 @@
+
+                    if (!records.isInitialized()) {
+                        try {
+                            fetchBuffer.setNextInLineFetch(initialize(records));
```
This is ok. However, it feels a bit unnatural since we take `records` from `fetchBuffer` only to call `fetchBuffer` to set the next in line. Could `fetchBuffer.peek()`, `initialize` and `setNextInLineFetch` be done in one method inside fetchBuffer?

kirktrue replied:
I absolutely agree that it's awkward. This is essentially copy-and-paste from the existing implementation and I wanted to change things as little as possible. I'll take another look and see if I can make it more intuitive and easy to follow.

kirktrue replied:
I'm gonna call a time out on this because I've spent a couple of hours looking into this and am getting nowhere.

My attempts are either a) subtly wrong because all tests fail!, or b) no more clear than the code is now.

It's an irritatingly impregnable bit of code, I'll admit. Do you mind if I file a Jira and come back to this a bit later?

LMK. Thanks!

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,143 @@
+            nextInLineFetch = null;
+    }
+
+    boolean maybeDrain(final Set<TopicPartition> partitions, final CompletedFetch completedFetch) {
```
Should this be private?

kirktrue replied:
Sure. Will change.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    Set<TopicPartition> partitions() {
```
partitions => fetchablePartitions?

kirktrue replied:
The partitions in the buffer are actually not the fetchable partitions, "by definition." When it comes time to send out fetch requests, if there is still data for a given partition in the buffer, we intentionally skip over it it as the user has not processed it yet.

I've added some documentation/comments around this area to hopefully make that clear.

LMK what you think.

junrao replied:
Perhaps bufferedPartitions?

kirktrue replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -98,16 +115,59 @@ public void onFailure(RuntimeException e) {
+            // we do not need to re-enable wakeups since we are closing already
+            client.disableWakeups();
+            maybeCloseFetchSessions(timer);
+            Utils.closeQuietly(decompressionBufferSupplier, "decompressionBufferSupplier");
```
`decompressionBufferSupplier` is in AbstractFetch. So, it feels a bit weird that it's being closed in `Fetcher`.

kirktrue replied:
I agree. I'll see what I can do.

kirktrue replied:
I pushed a change to clean up how the resources in `AbstractFetch` and `Fetcher` are released in the `close()` methods. The unit and integration tests pass (locally) and it _makes_ sense to me, but please LMK if there's anything that doesn't sit right with the change.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java:
```diff
@@ -0,0 +1,372 @@
+import java.util.Optional;
+import java.util.Queue;
+
+import static org.apache.kafka.clients.consumer.internals.AbstractFetch.requestMetadataUpdate;
```
It feels a bit weird for FetchCollector to depend on AbstractFetch.

kirktrue replied:
I went ahead and added `FetchUtils` back so that this utility method lives in a neutral third party.

kirktrue replied:
LMK if you're OK with this. Thanks!

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -63,6 +68,31 @@ public NetworkClientDelegate(
+        tryConnectNodes.add(node);
+    }
+
+    public void maybeTryConnect() {
```
This seems unused?

kirktrue replied:
Yes, it was. Thanks for catching this. It's now executed in the `DefaultBackgroundThread` right before we send out requests.

This is an artifact of manually merging different commits from the staging branch into a single PR for `trunk` 😞

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java:
```diff
@@ -0,0 +1,58 @@
+ * @see ConsumerNetworkClient
+ * @see NetworkClientDelegate
+ */
+public interface NodeStatusDetector {
```
Why do we need two implementations of NodeStatusDetector? They both seem to call NetworkClientUtils.

kirktrue replied:
One implementation is using the existing `ConsumerNetworkClient` which is closely related to the network/threading model of the existing `KafkaConsumer`. The newer implementation is `NetworkClientDelegate` which does not need the synchronization and odds-and-ends that `ConsumerNetworkClient` needs.

`NodeStatusDetector` gives us a general interface that a good chunk of the refactored code in `AbstractFetch` and elsewhere can use without having to tie us to a single implementation.

kirktrue replied:
LMK if you still have questions/concerns here.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -63,6 +68,31 @@ public NetworkClientDelegate(
+
+    @Override
+    public void tryConnect(Node node) {
+        tryConnectNodes.add(node);
```
Is this the correct implementation? It doesn't seem tryConnectNodes is being used.

kirktrue replied:
It is now 😉

Unlike the `ConsumerNetworkClient`, we don't attempt to connect to the node immediately. We queue up the set of `Node`s to which to connect and call them from the background thread shortly before calling `poll`. The existing implementation would attempt to connect immediately, and potentially more than once in some cases.

kirktrue replied:
LMK if that makes sense to you (or not). Thanks.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -67,35 +54,34 @@ public abstract class AbstractFetch<K, V> implements Closeable {
 
     public AbstractFetch(final LogContext logContext,
-                         final ConsumerNetworkClient client,
+                         NodeStatusDetector nodeStatusDetector,
```
Should this be final too?

kirktrue replied:
Yes. Fixed.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -205,14 +190,14 @@ protected void handleFetchResponse(final Node fetchTarget,
      *
      * @param fetchTarget {@link Node} from which the fetch data was requested
-     * @param e {@link RuntimeException} representing the error that resulted in the failure
+     * @param t {@link RuntimeException} representing the error that resulted in the failure
```
t is now a Throwable, not RuntimeException.

kirktrue replied:
Changed JavaDoc to match.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -50,7 +55,8 @@
     private final Logger log;
-    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private final FetchCollector<K, V> fetchCollector;
+    private final ConsumerNetworkClient client;
```
Hmm, why do we need to duplicate `client` in `Fetcher` and `AbstractFetch`? `client` is only used in `maybeCloseFetchSessions`. Could we move maybeCloseFetchSessions to AbstractFetch?

kirktrue replied:
I agree, it's ugly 😦

`client` and `nodeStatusDetector` both point at the same object in `Fetcher` and `AbstractFetch`, respectively, but the two classes use them for different purposes.

The `ConsumerNetworkClient` used in the `Fetcher` invokes methods like `poll()`, `send()`, and `disableWakeups()` to perform network I/O "directly." The forthcoming `FetchRequestManager` adheres to the design principle that request managers don't perform network I/O directly. To enforce that, we introduced this intentionally limited `NodeStatusDetector` to allow for checking basic state of nodes.

I'll keep looking at it a bit to see if I can come up with something cleaner because I don't like it either.

kirktrue replied:
I made a change here that I'm still not crazy about.

Instead of having a `NodeStatusDetector` instance variable in `AbstractFetch`, I introduced a type parameter in `AbstractFetch` that allows a subclass to provide a more specific type. Now `Fetcher` can provide the type as `ConsumerNetworkClient` and thus it can call the specialized methods on the instance variable from `AbstractFetch`.

The part I'm not crazy about is that the instance variable is still called `nodeStatusDetector`. I think I would be a bit confused when I see calls like `nodeStatusDetector.disableWakeups()`. But maybe that's just me. I thought of changing the name of the instance variable to `client` or `networkInfo` or other equally vague names.

If you have any input/suggestions/preferences, do LMK. Thanks!

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -405,7 +274,7 @@ private List<TopicPartition> fetchablePartitions() {
      * @param leaderReplica {@link Node} for the leader of the given partition
      * @param currentTimeMs Current time in milliseconds; used to determine if we're within the optional lease window
-     * @return Replica {@link Node node} from which to request the data
+     * @return Replic {@link Node node} from which to request the data
```
Replica

kirktrue replied:
Not sure how this keeps coming back 😦

kirktrue replied:
Fixed

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CommitApplicationEvent.java:
```diff
@@ -19,46 +19,33 @@
+        return "CommitApplicationEvent{" +
+                "offsets=" + offsets +
+                ", future=" + future +
+                ", type=" + type +
```
Could we restructure the `toString` code to avoid each event duplicating the shared fields in the parent class?

kirktrue replied:
Done.

kirktrue replied:
LMK if you're OK with the way I went about solving this. Thanks

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CompletedFetch.java:
```diff
@@ -253,7 +248,7 @@ private Record nextFetchedRecord() {
      * @return {@link ConsumerRecord Consumer records}
      */
-    List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
+    <K, V> List<ConsumerRecord<K, V>> fetchRecords(FetchConfig<K, V> fetchConfig, int maxRecords) {
```
Missing javadoc for `fetchConfig`.

kirktrue replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java:
```diff
@@ -0,0 +1,143 @@
+public class FetchBuffer implements Closeable {
+
+    private final Logger log;
+    private final ConcurrentLinkedQueue<CompletedFetch> completedFetches;
```
Could we add a comment that there is only up to one CompletedFetch per partition in `completedFetches`?

junrao replied:
Has this been addressed?

kirktrue replied:
I hadn't addressed it because I wasn't 100% sure if that was accurate. I temporarily added some assertion checking to verify that and re-run the unit tests and they passed, so I'll take that as sufficient confirmation. I've updated the class-level documentation to include that fact. LMK if you're OK with the comments. Thanks.

junrao reviewed (commented):
@kirktrue : Thanks for the PR. Made a pass of non testing files. Left a few comments.

kirktrue committed:
Updates based on feedback

kirktrue committed:
Update CompletedFetch.java

Added JavaDoc

kirktrue committed:
Update DefaultBackgroundThread.java

kirktrue committed:
Reinstituted FetchUtils to share functionality

kirktrue committed:
Added (hopefully) clarifying comments for fetchable partitions

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchUtils.java:
```diff
@@ -0,0 +1,53 @@
+    static void requestMetadataUpdate(final ConsumerMetadata metadata,
+                                      final SubscriptionState subscriptions,
+                                      final TopicPartition topicPartition) {
+        metadata.requestUpdate(true);
```
This was set to `false`. Why is it changed to `true`? I think `false` is correct since we don't want to reset the metadata retry count when receiving retriable error due to stale metadata.

kirktrue replied:
You're right. I double-checked the (very helpful) comments in `requestUpdate()` and I agree, so I set it to `false`.

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -256,138 +258,23 @@ protected FetchRequest.Builder createFetchRequest(final Node fetchTarget,
-     * </p>
+     * Return the list of <em>fetchable</em> partitions, which are the set of partitions to which we are subscribed,
+     * but <em>excluding</em> any partitions for which we still have buffered data. The idea is that since the user
+     * has yet to process the data for the partition that has already been fetch, we should not go send for more data
```
been fetch => been fetched

kirktrue replied:
Fixed. Also made this method return a `Set` which is more appropriate.

kirktrue committed:
Revised how AbstractFetch and Fetcher release resources when they close()

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -63,6 +68,31 @@ public NetworkClientDelegate(
+    }
+
+    @Override
+    public void tryConnect(Node node) {
```
`tryConnect` and `maybeTryConnect` seem highly coupled. Why is only the former part of the `NodeStatusDetector` interface? Also, it seems that we have used `maybeTryConnect`, but not `tryConnect` in this class.

kirktrue replied:
Yes—`tryConnect` and `maybeTryConnect` were intended to be used together. The reason the latter wasn't in the interface was because of the difference in _when_ we perform network I/O. In the existing `Consumer`, we perform it immediately, whereas in the refactored `Consumer`, we delay that I/O until the last step. So the intent was to for the new `Consumer` to queue up the connection attempts and perform them in one step in `maybeTryConnect` vs. immediately when `tryConnect` was called.

That said, a) it's confusing, and b) it's not needed in this particular PR, so I've removed the whole `NodeStatusDetector` for now.

kirktrue committed:
Updates to remove the duplicate client object in the Fetcher

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -91,23 +105,70 @@ public void onFailure(RuntimeException e) {
+        return fetchCollector.collectFetch(fetchBuffer);
+    }
+
+    protected void maybeCloseFetchSessions(final Timer timer) {
```
Hmm, I don't quite understand this refactoring now. This method only depends on `nodeStatusDetector`, which is in `AbstractFetch`. Why is this method moved here instead of remaining in `AbstractFetch`?

kirktrue replied:
Yes, it is a bit obtuse. My apologies.

There are two reasons that `maybeCloseFetchSessions` exists in `Fetcher`:

1. The forthcoming `FetchRequestManager` doesn't close sessions in the same way, so it doesn't make sense to have the method in the shared `AbstractFetch` class.
2. `FetchRequestManager` (as with the other `RequestManager`s) doesn't use the `ConsumerNetworkClient` but uses the `NetworkClientDelegate`. The code in `AbstractFetch` that is shared by both `Fetcher` and—soon—`FetchRequestManager` needed a way to call some of the methods that `ConsumerNetworkClient` and `NetworkClientDelegate` have in common. Thus the `NodeStatusDetector` interface was created.

Because the `FetchRequestManager` isn't in this PR, it's understandably unclear the reason for the current structure. For now I've made the following changes:

1. I moved `maybeCloseFetchSessions` to `AbstractFetch`
2. I left `AbstractFetch` with a reference to a `ConsumerNetworkClient`
3. I've removed the `NodeStatusDetector`

As needed, the above may be changed in forthcoming PRs.

kirktrue committed:
Merge branch 'trunk' into KAFKA-14274-2-introduce-new-classes

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NodeStatusDetector.java:
```diff
@@ -0,0 +1,58 @@
+    void maybeThrowAuthFailure(Node node);
+
+    /**
+     * Initiate a connection if currently possible. This is only really useful for resetting
```
Hmm, the implementation in ConsumerNetworkClient initiates a connection, but the implementation in NetworkClientDelegate doesn't. So, is this the right abstraction to capture both?

kirktrue replied:
Per the above comments, I've removed `NodeStatusDetector` for now, so this should be moot for the time being.

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. Made another pass of non testing files. A few more comments.

kirktrue committed:
Renamed FetchBuffer.partitions() to FetchBuffer.bufferedPartitions()

kirktrue committed:
Merge branch 'KAFKA-14274-2-introduce-new-classes' of github.com:kirktrue/kafka into KAFKA-14274-2-introduce-new-classes

kirktrue committed:
Update FetchBuffer.java

Updated comments

kirktrue committed:
Update CompletableApplicationEvent.java

Fixed merge conflict error.

kirktrue committed:
Update FetchUtils.java

Set the requestUpdate parameter to false since we're requesting an update due to leader issues vs. adding new topics.

kirktrue committed:
Updated fetchablePartitions to return a Set, not List

And fixed typo in comments

kirktrue committed:
Backing out of introducing the NetworkClientDelegate

kirktrue committed:
Update AbstractFetch.java

Two more changes to revert the parameter to metadata.requestUpdate() to be false instead of true

junrao commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -421,14 +309,45 @@ Node selectReadReplica(final TopicPartition partition, final Node leaderReplica,
         }
     }
 
+    protected Map<Node, FetchSessionHandler.FetchRequestData> prepareCloseFetchSessionRequests() {
```
This is only used inside AbstractFetch. Could this be private?

kirktrue replied:
Done

junrao commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/CompletedFetchTest.java:
```diff
@@ -115,13 +122,17 @@ public void testCommittedTransactionRecordsIncluded() {
-                OffsetResetStrategy.NONE,
-                true,
-                0,
+        CompletedFetch completedFetch = newCompletedFetch(0,
```
Could we merge the next line in? Ditto for similar places below.

kirktrue replied:
Done. I double-checked that I caught the places where the method args were unnecessarily broken over two lines, but let me know if I missed any. Thanks!

junrao commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchBufferTest.java:
```diff
@@ -0,0 +1,193 @@
+ * internally:
+ *
+ * <ol>
+ *     <li>A special "next in line" buffer</li>
```
Are these comments relevant to this class?

kirktrue replied:
Yes, but apparently, they're confusing, though 😄

I tried to reword the comments, but I'm not sure if it was enough. The idea was to emphasize that the tests are trying to ensure that we keep the internal state that is split between the main buffer queue and the 'next in line' buffer consistent.

junrao commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchBufferTest.java:
```diff
@@ -0,0 +1,193 @@
+     * Tests that the buffer returns partitions for both the queue and the next-in-line buffer.
+     */
+    @Test
+    public void testPartitions() {
```
testBufferedPartitions?

kirktrue replied:
Done.

junrao commented on clients/src/main/java/org/apache/kafka/common/internals/IdempotentCloser.java:
```diff
@@ -0,0 +1,127 @@
+        this.isClosed = new AtomicBoolean(isClosed);
+    }
+
+    public void maybeThrowIllegalStateException(Supplier<String> message) {
```
Could we describe what the method does?

kirktrue replied:
Yes. I added comments to the methods in this class. LMK if they there's additional detail needed. Thanks!

junrao replied:
The comment makes sense. Would `assertOpen` be a clearer method name?

kirktrue replied:
Agreed!

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. Made a pass of most files. A few more comments.

kirktrue committed:
Changed AbstractFetch.prepareCloseFetchSessionRequests to be private

kirktrue committed:
IdempotentCloser JavaDoc changes

kirktrue committed:
Minor formatting changes to CompletedFetchTest

kirktrue committed:
Updated FetchBufferTest.testPartitions() to testBufferedPartitions() and more changes to comments

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. Just one minor comment. Also, it seems the PR can't be built.

kirktrue committed:
Removed superfluous imports that caused checkstyle failures

kirktrue committed:
Renamed maybeThrowIllegalStateException to assertOpen

junrao reviewed (commented):
@kirktrue : Thanks for the updated PR. LGTM. 45 test failures. Are they related to this PR?

kirktrue commented:
@junrao yeah, lots of errors in the socket/selector level for some reason:

```
kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeWithWildcardAcls()
kafka.api.TransactionsTest.testBumpTransactionalEpoch()
kafka.network.SocketServerTest.testControlPlaneAsPrivilegedListener()
kafka.network.SocketServerTest.testControlPlaneTakePrecedenceOverInterBrokerListenerAsPrivilegedListener()
kafka.network.SocketServerTest.testInterBrokerListenerAsPrivilegedListener()
kafka.network.SocketServerTest.testStagedListenerShutdownWhenConnectionQueueIsFull()
kafka.network.SocketServerTest.testStagedListenerStartup()
kafka.network.SocketServerTest.testZeroMaxConnectionsPerIp()
kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestExcludingClusterAuthorizedOperations()
kafka.server.DescribeClusterRequestTest.testDescribeClusterRequestIncludingClusterAuthorizedOperations()
o.a.k.common.network.SelectorTest.testConnectionsByClientMetric()
o.a.k.common.network.SelectorTest.testInboundConnectionsCountInConnectionCreationMetric()
o.a.k.common.network.SelectorTest.testMuteOnOOM()
o.a.k.common.network.Tls12SelectorTest.testConnectionsByClientMetric()
o.a.k.common.network.Tls12SelectorTest.testInboundConnectionsCountInConnectionCreationMetric()
o.a.k.common.network.Tls12SelectorTest.testMuteOnOOM()
o.a.k.common.network.Tls13SelectorTest.testConnectionsByClientMetric()
o.a.k.common.network.Tls13SelectorTest.testInboundConnectionsCountInConnectionCreationMetric()
o.a.k.common.network.Tls13SelectorTest.testMuteOnOOM()
o.a.k.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector()
o.a.k.controller.QuorumControllerTest.testBalancePartitionLeaders()
o.a.k.image.loader.MetadataLoaderTest.testNoPublishEmptyImage()
o.a.k.streams.integration.NamedTopologyIntegrationTest.shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing()
o.a.k.tiered.storage.integration.TransactionsWithTieredStoreTest.testAbortTransactionTimeout()
o.a.k.tiered.storage.integration.TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout()
o.a.k.tools.JmxToolTest.initializationError()
```

I don't think any of these are related, but I'm going to force a rebuild, just to check.

junrao reviewed (approved):
@kirktrue : The 7 test failures seem unrelated. The PR LGTM.

chia7712 commented:
```
        // Return all partitions that are in an otherwise fetchable state *and* for which we don't already have some
        // messages sitting in our buffer.
        return new HashSet<>(subscriptions.fetchablePartitions(isNotBuffered));
```

Out of curiosity, why are they being copied into a new `Set` collection? The `ShareConsumeRequestManager` directly use the `List` type without an extra deep copy. Also, the elements returned by `subscriptions.fetchablePartitions` should already be deduplicated.

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java#L1076
